pennylane/capture/primitives.py

```diff
@@ -0,0 +1,119 @@
+# Copyright 2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This submodule defines the abstract classes and primitives for capture.
+"""
+
+from functools import lru_cache
+from typing import Optional
+
+import pennylane as qml
+
+has_jax = True
+try:
+    import jax
+except ImportError:
+    has_jax = False
+
+
+@lru_cache  # construct the first time lazily
+def _get_abstract_operator() -> type:
+    """Create an AbstractOperator once in a way protected from lack of a jax install."""
+    if not has_jax:  # pragma: no cover
+        raise ImportError("Jax is required for plxpr.")  # pragma: no cover
+
+    class AbstractOperator(jax.core.AbstractValue):
+        """An operator captured into plxpr."""
+
+        # pylint: disable=missing-function-docstring
+        def at_least_vspace(self):
+            # TODO: investigate the proper definition of this method
+            raise NotImplementedError
+
+        # pylint: disable=missing-function-docstring
+        def join(self, other):
+            # TODO: investigate the proper definition of this method
+            raise NotImplementedError
+
+        # pylint: disable=missing-function-docstring
+        def update(self, **kwargs):
+            # TODO: investigate the proper definition of this method
+            raise NotImplementedError
+
+        def __eq__(self, other):
+            return isinstance(other, AbstractOperator)
```

Comment:
Honestly I was just kinda duplicating this from the prototype.  In the measurement PR #5564  I gave `AbstractMeasurement` properties for `n_wires` and an `abstract_eval` function, since that information was needed later in the pipeline to determine shapes.

I'm not sure if `num_qubits` will really work, since for things like `X(a) + X(b)`, we can't really determine how many wires there are without knowing the concrete wire values.

I'd be happy to extend with additional metadata if we see a need.

Reply:
> I'm not sure if num_qubits will really work, since for things like X(a) + X(b), we can't really determine how many wires there are without knowing the concrete wire values.

Hmm, so there are cases where we would need the unique (non-overlapping) set of wires from a set of operations? That is tricky indeed, we might even this problem in Catalyst already 🤔

Reply:
> I'd be happy to extend with additional metadata if we see a need.

Fair enough 👍

Reply:
@dime10 yeah I recall this came up as an issue a few weeks ago, [maybe this one](https://xanaduhq.slack.com/archives/C06CNPQLK2T/p1713562876925589)?

E.g.,

```python
def circuit(w1, w2):
    H = qml.X(w1) + qml.Y(w2)
    return qml.probs(op=H)
```